add model scaling section

lvwerra opened the pull request:
# What does this PR do?

This PR adds a practical scaling guide to the documentation. When starting to work with large transformer models one of the most common issues is running out of GPU memory. There are several easy to implement strategies to counteract this. This guide shows how to use gradient accumulation, gradient checkpointing, mixed precision training, and choice of optimizer to decrease the memory footprint. It shows how to do this with both the `Trainer` and `accelerate` in a hands-on fashion.

## What does this PR not do?
This PR does **not** replace the in-depth explanations at ["Model Parallelism"](https://huggingface.co/docs/transformers/parallelism) and ["Performance and Scalability: How To Fit a Bigger Model and Train It Faster"](https://huggingface.co/docs/transformers/performance). Rather the proposed guide should be an easy entry point for new users whereas these other sections are for more advanced users.

## TensorFlow?
If this guide resonates well it would make sense to extend it with a TensorFlow/Keras section (similar to the `accelerate` section). Maybe TensorFlow boy (aka @Rocketknight1) has some ideas?

Commit:
add model scaling section

Rocketknight1 commented:
We could definitely make a TF/Keras section! I believe Keras doesn't support gradient accumulation or gradient checkpointing out of the box, though, but mixed precision, Adafactor, etc. are all very doable.

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+}
+```
+
+> Note: In order to properly clear the memory after experiments we need restart the Python kernel between experiments. Run all steps above and then just one of the experiments below.
````
Use the <Tip> syntax here for better rendering.

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+## Vanilla Training
+
+As a first experiment we will use the `Trainer` and train the model without any further modifications and a batch size of 4:
```
```suggestion
As a first experiment we will use the [`Trainer`] and train the model without any further modifications and a batch size of 4:
```

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+
+As a first experiment we will use the `Trainer` and train the model without any further modifications and a batch size of 4:
+
+```py
+training_args = TrainingArguments(per_device_train_batch_size=4, **default_args)
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 57.82
+    Samples/second: 8.86
+    GPU memory occupied: 14949 MB.
````
Use the doctest syntax or put the result in a code block.

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+The idea behind gradient accumulation is to instead of calculating the gradients for the whole batch at once to do it in smaller steps. The way we do that is to calculate the gradients iteratively in smaller batches by doing a forward and backward pass through the model and accumulating the gradients in the process. When enough gradients are accumulated we run the model's optimization step. This  ...
+
+We can use gradient accumulation in the `Trainer` by simply adding the `gradient_accumulation_steps` argument to `TrainingArguments`. Let's see how it impacts the models memory footprint:
```
```suggestion
We can use gradient accumulation in the [`Trainer`] by simply adding the `gradient_accumulation_steps` argument to [`TrainingArguments`]. Let's see how it impacts the models memory footprint:
```

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+## Gradient Accumulation
+
+The idea behind gradient accumulation is to instead of calculating the gradients for the whole batch at once to do it in smaller steps. The way we do that is to calculate the gradients iteratively in smaller batches by doing a forward and backward pass through the model and accumulating the gradients in the process. When enough gradients are accumulated we run the model's optimization step. This  ...
```
It's not obvious from the intro that this is just a way to train at a bigger batch size while not going OOM.

lvwerra replied:
I modified the last sentence to reflect that better.

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+
+We can use gradient accumulation in the `Trainer` by simply adding the `gradient_accumulation_steps` argument to `TrainingArguments`. Let's see how it impacts the models memory footprint:
+
+```py
+training_args = TrainingArguments(per_device_train_batch_size=1, gradient_accumulation_steps=4, **default_args)
+
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 66.03
+    Samples/second: 7.75
+    GPU memory occupied: 8681 MB.
````
Same remark as above.

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+Gradient checkpointing strikes a compromise between the two approaches and saves strategically selected activations throughout the computational graph so only a fraction of the activations need to be re-computed for the gradients. See [this great article](https://medium.com/tensorflow/fitting-larger-networks-into-memory-583e3c758ff9) explaining the ideas behind gradient checkpointing. 
+
+To enable gradient checkpointing in the `Trainer` we only need ot pass it as a flag to the `TrainingArguments`. Everything else is handled under the hood:
```
```suggestion
To enable gradient checkpointing in the [`Trainer`] we only need ot pass it as a flag to the [`TrainingArguments`]. Everything else is handled under the hood:
```

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+print_summary(result)
+```
+
+    Time: 85.47
+    Samples/second: 5.99
+    GPU memory occupied: 6775 MB.
````
Same remark as above.

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+## FP16 Training
+
+The idea of mixed precision training is that no all variables need to be stored in full (32-bit) floating point precision. If we can reduce the precision the variales and their computations are faster and use less memory. The main advantage comes from saving the activations in half (16-bit) precision. Although the gradients are also computed in half precision they are converted back to full preci ...
```
No, fp16 actually results in more memory usage in most cases when batch size is small and the model is big, because there is 1.5 x the size of the model on the GPU (one copy in FP32 and one copy in FP16).

This is a common source of confusion for users.

lvwerra replied:
including me :) i fixed the text.

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+    GPU memory occupied: 13939 MB.
+
+
+We can see that this is almost twice as fast as the vanilla training while also requiring less GPU memory. Let's add it to the mix of the previous methods:
```
I would remove the "requiring less GPU memory" as it's only a few MBs.

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+print_summary(result)
+```
+
+    Time: 50.76
+    Samples/second: 10.09
+    GPU memory occupied: 7275 MB.
````
Same remark as above.

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+print_summary(result)
+```
+
+    Time: 64.31
+    Samples/second: 7.96
+    GPU memory occupied: 12295 MB.
````
Same remark as above.

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+print_summary(result)
+```
+
+    Time: 56.54
+    Samples/second: 9.06
+    GPU memory occupied: 4847 MB.
````
Same remark as above.

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+Instead of aggregating optimizer states like Adafactor, 8-bit Adam keeps the full state and quantizes it. Quantization means that it stores the state with lower precision and dequantizes it only for the optimization. This is similar to the idea behind FP16 training where using variables with lower precision saves memory. 
+
+In contrast to the previous approaches is this one not integrated into the `Trainer` as a simple flag. We need to install the 8-bit optimizer and then pass it as a custom optimizer to the `Trainer`. Follow the installation guide in the Github [repo](https://github.com/facebookresearch/bitsandbytes) to install the `bitsandbytes` library that implements the 8-bit Adam optimizer.
```
```suggestion
In contrast to the previous approaches is this one not integrated into the [`Trainer`] as a simple flag. We need to install the 8-bit optimizer and then pass it as a custom optimizer to the [`Trainer`]. Follow the installation guide in the Github [repo](https://github.com/facebookresearch/bitsandbytes) to install the `bitsandbytes` library that implements the 8-bit Adam optimizer.
```
+ link to the bitsandbytes GitHub.

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+print_summary(result)
+```
+
+    Time: 55.95
+    Samples/second: 9.15
+    GPU memory occupied: 13085 MB.
````
Same remark as above.

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+    },
+]
+
+optimizer_kwargs = {
+    "betas": (training_args.adam_beta1, training_args.adam_beta2),
+    "eps": training_args.adam_epsilon,
+}
+optimizer_kwargs["lr"] = training_args.learning_rate
+adam_bnb_optim = bnb.optim.Adam8bit(optimizer_grouped_parameters, **optimizer_kwargs)
```
Can we just pass the kwargs when calling `Adam8bit`? I don't see why we store them in a dict.

lvwerra replied:
Good point, a relic from copy-pasting.

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+print_summary(result)
+```
+
+    Time: 49.46
+    Samples/second: 10.35
+    GPU memory occupied: 5363 MB.
````
Same remark as above.

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+    
+![png](https://huggingface.co/datasets/lvwerra/repo-images/raw/main/gpu-memory-savings.png)
+
+## Using `accelerate`
```
```suggestion
## Using 🤗 Accelerate
```

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+## Using `accelerate`
+
+So far we have used the `Trainer` to run the experiments but a more flexible alternative to that approach is to use [`accelerate`](https://huggingface.co/docs/accelerate/). With `accelerate` you have full control over the training loop and can essentially write the loop in pure PyTorch with some minor modifications. In turn it allows you to easily scale across different infrastructures such as CP ...
```
```suggestion
So far we have used the [`Trainer`] to run the experiments but a more flexible alternative to that approach is to use [`accelerate`](https://huggingface.co/docs/accelerate/). With 🤗 Accelerate you have full control over the training loop and can essentially write the loop in pure PyTorch with some minor modifications. In turn it allows you to easily scale across different infrastructures such as CPUs, GPUs, TPUs, or distributed multi-GPU setups without changing any code. Let's see what it takes to implement all of the above tweaks in accelerate. We can still use the [`TrainingArguments`] to wrap the training settings:
```

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+)
+```
+
+The full example training loop with `accelerate` is only a handful of lines of code long: 
````
```suggestion
The full example training loop with 🤗 Accelerate is only a handful of lines of code long: 
```

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+        optimizer.zero_grad()
+```
+
+First we wrap the dataset in a `DataLoader`. Then we can enable gradient checkpointing by calling the model's `gradient_checkpointing_enable` method. When we initialize the `Accelerator` we can specifiy if we want to use mixed precision training and it will take care of it for us in the `prepare` call. During the `prepare` call the dataloader will also be distributed across workers should we use  ...
````
Linking to the doc of all those objects would be great!

sgugger commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+Finally, we can write the main training loop. Note that the `backward` call is handled by accelerate. We can also see how gradient accumulation works: we normalize the loss so we get the average at the end of accumulation and once we have enough steps we run the optimization. Now the question is: does this use the same amount of memory as the previous steps? Let's check:
+
+
+```py
+print_gpu_utilization()
+```
+
+    GPU memory occupied: 5363 MB.
````
Use the doctest syntax here.

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+    GPU memory occupied: 5363 MB.
+
+
+Indeed it does. Implementing these optimization techniques with `accelerate` only takes a handful of lines of code and comes with the benefit of more flexiblity in the training loop.
```
```suggestion
Indeed it does. Implementing these optimization techniques with 🤗 Accelerate only takes a handful of lines of code and comes with the benefit of more flexiblity in the training loop.
```

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+If your model fits on a single GPU scaling to many GPUs can be achieved fairly easily with data parallelism. The idea is very similar to gradient accumulation with the distinction that instead of running the forward and backward passes during the accumulation in sequence on a single machine they are performed in parallel on multiple machines. So each GPU gets a small batch, runs the forward and b ...
+
+The `Trainer` allows for distributed training and if you execute your `Trainer` training script on a machine with multiple GPUs it will automatically utilize all of them, hence the name `per_device_train_batch_size`. In `accelerate` you can configure the infrastructure setup with the following command:
```
```suggestion
The [`Trainer`] allows for distributed training and if you execute your [`Trainer`] training script on a machine with multiple GPUs it will automatically utilize all of them, hence the name `per_device_train_batch_size`. In 🤗 Accelerate you can configure the infrastructure setup with the following command:
```

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+## What if my model still does not fit?
+
+If the model does not fit on a single GPU with all the mentioned tricks there are still more methods we can apply although life starts to get a bit more complicated. This usually involves some form of pipeline or tensor parallelism where the model itself is distributed across several GPUs. One can also make use of DeepSpeed which implements some of these parallelism strategies along with some mor ...
```
```suggestion
If the model does not fit on a single GPU with all the mentioned tricks there are still more methods we can apply although life starts to get a bit more complicated. This usually involves some form of pipeline or tensor parallelism where the model itself is distributed across several GPUs. One can also make use of DeepSpeed which implements some of these parallelism strategies along with some more optimization to reduce the memory footprint such as partitioning the optimizer states. You can read more about this in the ["Model Parallelism" section](parallelism). 
```

sgugger commented on docs/source/scaling.mdx:
```diff
@@ -0,0 +1,446 @@
+
+If the model does not fit on a single GPU with all the mentioned tricks there are still more methods we can apply although life starts to get a bit more complicated. This usually involves some form of pipeline or tensor parallelism where the model itself is distributed across several GPUs. One can also make use of DeepSpeed which implements some of these parallelism strategies along with some mor ...
+
+This concludes this guide for scaling the training of large models. There is a lot more going on under the hood than we looked at and the ["Performance and Scalability: How To Fit a Bigger Model and Train It Faster" section](https://huggingface.co/docs/transformers/performance) goes into much more detail about the introduced optimization techniques we introduced.
```
```suggestion
This concludes this guide for scaling the training of large models. There is a lot more going on under the hood than we looked at and the ["Performance and Scalability: How To Fit a Bigger Model and Train It Faster" section](performance) goes into much more detail about the introduced optimization techniques we introduced.
```

sgugger reviewed (approved):
Thanks a lot for working on this, this is a great addition!

The way you show the outputs is inconsistent with the rest of the doc. It should either be with the doctest syntax:
```
>>> instruction
result 
```
or in two separate code blocks.

lvwerra commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,446 @@
+The `nvidia-ml-py3` library allows us to monitor the memory usage of the models from within Python. You might be familiar with the `nvidia-smi` command in the terminal - this library allows to access the same information in Python directly. Let's import it via `pynvml` along with a few other libraries that we'll use:
+
+
+```py
+from pynvml import *
+import numpy as np
+
+from datasets import Dataset
+from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer, logging
+
+logging.set_verbosity_error()
````
Only the first time they appear or every time? E.g. the `Trainer` is used all over and in order to make them individually copy-pastable I'd have to import it in every cell, right?

sgugger replied:
Only the first time is fine!

lvwerra committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Commit:
integrate reviewer feedback

lvwerra commented:
Hi @sgugger I integrated your feedback, let me know if this is what you had in mind.

stas00 commented on docs/source/scaling.mdx:
````diff
@@ -0,0 +1,454 @@
+```
+
+That looks good: the GPU memory is not occupied as we would expect before we load any models. If that's not the case on your machine make sure to stop all processes that are using GPU memory.
+
````
at this point I typically do `torch.ones((1, 1)).to(device_id)` because cuda kernels take up 1-2GB of each card.

https://github.com/stas00/ipyexperiments/blob/39c9b454e89e53b74c74dcb579c12ecf3d2161b9/ipyexperiments/utils/mem.py#L21-L28

Because not all of card's memory is available to the user, but once it's loaded the user knows how much memory is really available.

stas00 reviewed (commented):
Thanks for working on it, @lvwerra. We need more of similar content.

That's said I think a better solution would be to rewrite the existing documents and not add a new one that has more than 50% overlap with the existing doc, which will make it more confusing to the user and more difficult to maintain for us, as many things will now be duplicated if this PR is merged as is.

Indeed the original performance doc has been grown into scalability+performance and parallelism is no longer parallelism but actually it's the scalability doc, that's specifically there to deal with OOM.

I think we need 2 docs:

1. scalability
2. performance

and move/combine this PR's doc with the scalability sections of the the current performance doc - removing the huge overlap that this PR proposes.

and for the performance doc leave only software/hardware bits that talk about speeding things up, and minimally talking about memory usage.

Bottom line, IMHO it'd be good to brainstorm how to do that in the most harmonious way before creating a new document.

sgugger commented:
In the case of documentation, I don't think repetition is a bad thing since users rarely look at every page in a row. Making sure the docs are all linked to each other for further reading is important. The two doc pages reach different target audience IMO and present the techniques in different ways, so I think having both of them is good.

stas00 commented:
Repetition is a bad thing in this situation, IMHO:

1. as it'd make it more difficult to point users to the right doc. For example @LysandreJik has been pointing users to the performance doc when they have difficulties with getting their model to run memory-wise. Now you will have 2 competing docs, which one should he point users to? It shouldn't take more than a few secs for a maintainer to find where to point the user to and not needing to reread each doc and contemplate is it A or B?
2. how do you maintain 2 overlapping documents?
3. some users do read docs and now they will need to read 2 docs with confusing overlaps

But the big question is this why make things more complicated and messy when the whole performance/scalability document is due to a revamp anyway. It's far from great at the moment. I propose to discuss how we best serve both users and ourselves by improving the performance/scalability doc structure/layout. I'm not proposing to dropping anything from the current PR but to integrate rather than creating a diverging doc.

stas00 commented:
update: @lvwerra and I will discuss a revamp/integration on Fri.

Commit:
initialize GPU properly

Commit:
add note about BnB optimizer

lvwerra commented:
Hi there, I updated the doc and this is ready for another review:
- use @stas00's suggestion to initialize the measurement by loading a small tensor to the GPU
- added a note about the embedding issue with BnB optimizer

I disusssed with @stas00 that we could start the revamp by merging this PR and then move things around as we start adding the other documents in #15213.

stas00 reviewed (approved):
Looks great, @lvwerra 

I just wonder whether for now we should append this to the end of `performance.mdx`? Since we don't intend to have a `scaling.mdx` in the revamped docs - at least not in the initial layout.

otherwise if people start linking to `scaling.mdx` and then we wipe it out - we will need to deal with additional redirects or end up with broken links.

or at the very least putting a big fat disclaimer that this doc is temporary and will be removed soon. do not link to it.

Commit:
move doc from `scaling.mdx` to `performance.mdx`

lvwerra commented:
Good point, I merged `scaling.mdx` and `performance.mdx`. The document starts with the new practical guide and I added the existing material to a section called `Further discussions`. There is still a bit of redundancy between the two parts but I think we can refactor it when we flesh out the other parts.

@sgugger are you happy with the current state to be merged?

sgugger commented on docs/source/_toctree.yml:
```diff
@@ -25,6 +25,7 @@
     title: Summary of the tokenizers
   - local: multilingual
     title: Multi-lingual models
+
```
Not sure this new line is intentional. It shouldn't be needed.

lvwerra replied:
Not sure where that came from - deleted.

sgugger commented on docs/source/performance.mdx:
```diff
@@ -15,10 +15,466 @@ limitations under the License.
 -->
 
 # Performance and Scalability: How To Fit a Bigger Model and Train It Faster
+> _Or how to escape the dreaded "RuntimeError: CUDA error: out of memory" error._
```
A new line here would be nice for readability.

lvwerra replied:
added

sgugger commented on docs/source/performance.mdx:
````diff
@@ -15,10 +15,466 @@ limitations under the License.
+GPU memory occupied: 8681 MB.
+```
+
+We can see that the memory footprint was dramatically reduced at the cost of being only slightly slower than the vanilla run. Of course, this would change as you increase the number of accumulation steps. In general you would want to max out the GPU usage as much as possible. So in our case, the batch_size of 4 was already pretty close to the GPU's limit. If we wanted to train with a batch size o ...
````
Very nice remark!

sgugger commented on docs/source/performance.mdx:
````diff
@@ -15,10 +15,466 @@ limitations under the License.
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+```
````
Adding a new line here for readability would be nice.

lvwerra replied:
added a newline between every input/output block in the document

sgugger commented on docs/source/performance.mdx:
````diff
@@ -15,10 +15,466 @@ limitations under the License.
+
+### Adafactor
+
+Instead of keeping the rolling average for each element in the weight matrices Adafactor only stores aggregated information (row- and column-wise sums of the rolling averages) which reduces the footprint considerably. One downside of Adafactor is that in some instances convergence can be slower than Adam's so some experimentation is advised here. We can use Adafactor simply by setting `adafactor= ...
+
+
+```py
+training_args = TrainingArguments(per_device_train_batch_size=4, adafactor=True, **default_args)
````
I think this way is deprecated and that one has to pass optim-"adafactor" nowadays. @stas00 can confirm.

lvwerra replied:
Indeed - fixed.

sgugger commented on docs/source/performance.mdx:
```diff
@@ -15,10 +15,466 @@ limitations under the License.
+    gradient_accumulation_steps=4,
+    gradient_checkpointing=True,
+    fp16=True,
+    adafactor=True,
```
Same here.

sgugger reviewed (approved):
LGTM, thanks for iterating on this!

Commit:
integrate reviewer feedback

stas00 commented:
Thanks, @lvwerra 

1. Please revert all the header level changes in the existing doc - the menu on the right only handles ## and ### and your proposed change will practically eliminate most of the menu entries as you added an additional level

2. We discussed adding this PR's content at the end of the existing document as at the moment it's the extra and not the main document as it was laid out. 95% of this PR's material will be spread out over the new documents, and the current performance doc will remain here as a reference that the more specific docs will point to for indepth information. (e.g. mixed precision section we aren't going to repeat the full discussion about why and how in each sub-document and will point to the original instead). 

Does it make sense? If not, and we need to sync on the big picture - let's discuss it then first.

lvwerra commented:
Hi @stas00 

Sure, I'll revert the headers, I was not aware of that limitation.

Regarding the order of the content: I know that we will change that document radically in later PRs but I thought since this is gonna be public and maybe here for a few weeks it would be nicer to have the practical guide at the beginning and the theoretical part at the end. If you feel strongly about this I can change it back.

stas00 commented:
> Hi @stas00
> 
> Sure, I'll revert the headers, I was not aware of that limitation.

In fact I asked when the new design was added to support more than 3-levels but nothing came out of that. i.e. the menu is missing already a bunch of entries where there is a deeper nesting like the deepspeed doc. :(

> Regarding the order of the content: I know that we will change that document radically in later PRs but I thought since this is gonna be public and maybe here for a few weeks it would be nicer to have the practical guide at the beginning and the theoretical part at the end. If you feel strongly about this I can change it back.

Sure, that's fine. I think what I will do instead in the new PR is to move this original document out to a new document.

Commit:
revert section levels

LysandreJik commented:
Thanks a lot for working on this important part of the docs!
